returnn/datasets/concat_files.py

```diff
@@ -0,0 +1,613 @@
+        self.parent_conn.send(("init_seq_order", {"epoch": self.full_epoch_0idx + 1}))
+        self._num_seqs: Optional[int] = None
+
+    def _lazy_wait_for_init_seq_order(self):
+        if self._num_seqs is not None:
+            return
+        msg, num_seqs = self.parent_conn.recv()
+        assert msg == "num_seqs" and isinstance(num_seqs, int)
+        self._num_seqs = num_seqs
+
+    def get_num_seqs(self) -> int:
+        """num seqs for this sub epoch"""
+        self._lazy_wait_for_init_seq_order()
+        return self._num_seqs
+
+    def get_data_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        """get data seq"""
+        self._lazy_wait_for_init_seq_order()
+        self.parent_conn.send(("get_data_seq", {"seq_idx": seq_idx}))
+        msg, data = self.parent_conn.recv()
+        assert msg == "data_seq"
+        return data
+
+    def exit(self, *, join: bool = True):
+        """exit"""
+        self._lazy_wait_for_init_seq_order()
+        self.parent_conn.send(("exit", {}))
+        if join:
+            self.worker_proc.join()
+        if self.exit_hook:
+            self.exit_hook()
+            self.exit_hook = None
+
+    def __del__(self):
+        # noinspection PyBroadException
+        try:
+            self.exit(join=False)
+        except Exception:
+            pass
+        else:
+            try_run(self.worker_proc.join)
+
+
+def _worker_proc_loop(
+    epoch: int,
+    buffer_size: int,
+    dataset_dict: Dict[str, Any],
+    parent_conn: mpConnection,
+):
+    if sys.platform == "linux":
+        with open("/proc/self/comm", "w") as f:
+            f.write(f"CFD worker {epoch}")
+
+    assert isinstance(epoch, int) and isinstance(buffer_size, int)
+    assert isinstance(dataset_dict, dict)
+    assert isinstance(parent_conn, mpConnection)
+
+    dataset = init_dataset(dataset_dict)
+
+    got_init_seq_order = False
+    cache: List[DatasetSeq] = []
+    next_seq_idx = 0
+
+    # noinspection PyShadowingNames
+    def _add_to_cache():
+        nonlocal next_seq_idx
+        if len(cache) >= buffer_size:
+            return False
+        if not dataset.is_less_than_num_seqs(next_seq_idx):
+            return False
+        dataset.load_seqs(next_seq_idx, next_seq_idx + 1)
+        seq_tag = dataset.get_tag(next_seq_idx)
+        features = {data_key: dataset.get_data(next_seq_idx, data_key) for data_key in dataset.get_data_keys()}
+        res = DatasetSeq(seq_idx=next_seq_idx, seq_tag=seq_tag, features=features)
+        cache.append(res)
+        next_seq_idx += 1
+        return True
+
+    # noinspection PyShadowingNames
+    def _get_from_cache(seq_idx: int) -> Optional[DatasetSeq]:
+        if not cache:
+            return None
+        if seq_idx > cache[-1].seq_idx:
+            return None
+        for seq in cache:
+            assert seq.seq_idx <= seq_idx
+            if seq.seq_idx == seq_idx:
+                return seq
```

Comment:
This linear search seems slow if there are many seqs in the file? Would it make sense to make the cache a dict on the seq_idx instead?

Reply:
The buffer size is 1 by default, and in most cases should not be so large (not more than 10 or so). A dict is convenient here for getting, but for emptying part of the buffer, it would even require a scan through all elements, so even less efficient.